tests/infrastructure/src/test/java/org/mule/test/infrastructure/SpanTestHierarchyTestCase.java

```diff
@@ -80,6 +81,118 @@ public class SpanTestHierarchyTestCase extends AbstractMuleTestCase {
+    List<CapturedExportedSpan> capturedExportedSpans = getCapturedExportedSpansWithTraceState();
+
+    SpanTestHierarchy spanTestHierarchy = new SpanTestHierarchy(capturedExportedSpans);
+    spanTestHierarchy.withRoot(EXPECTED_FLOW_SPAN_NAME)
+        .addTraceStateKeyPresentAssertion("key1")
+        .addTraceStateKeyPresentAssertion("key2")
+        .addTraceStateKeyPresentAssertion("key3")
+        .beginChildren()
+        .child(EXPECTED_ASYNC_SPAN_NAME)
+        .addTraceStateKeyPresentAssertion("key1")
+        .addTraceStateKeyPresentAssertion("key2")
+        .addTraceStateKeyPresentAssertion("key3")
+        .beginChildren()
+        .child(EXPECTED_LOGGER_SPAN_NAME)
+        .addTraceStateKeyPresentAssertion("key4")
+        .endChildren()
+        .endChildren();
+
+    spanTestHierarchy.assertSpanTree();
+  }
+
+  @Test
+  public void testWhenTraceStateKeyNonExistedAssertedAndPresentAssertionShouldFail() {
+    expectedException.expectMessage("The span mule:logger has trace state key key3 and it must not be present");
+    List<CapturedExportedSpan> capturedExportedSpans = getCapturedExportedSpansWithTraceState();
+
+    SpanTestHierarchy spanTestHierarchy = new SpanTestHierarchy(capturedExportedSpans);
+    spanTestHierarchy.withRoot(EXPECTED_FLOW_SPAN_NAME)
+        .addTraceStateKeyPresentAssertion("key1")
+        .addTraceStateKeyPresentAssertion("key2")
+        .addTraceStateKeyPresentAssertion("key3")
+        .beginChildren()
+        .child(EXPECTED_ASYNC_SPAN_NAME)
+        .addTraceStateKeyPresentAssertion("key1")
+        .addTraceStateKeyPresentAssertion("key2")
+        .addTraceStateKeyPresentAssertion("key3")
+        .beginChildren()
+        .child(EXPECTED_LOGGER_SPAN_NAME)
+        .addTraceStateKeyNotPresentAssertion("key3")
+        .endChildren()
+        .endChildren();
+
+    spanTestHierarchy.assertSpanTree();
+  }
+
+  @Test
+  public void testAllTraceStateKeyValuesPresent() {
+    List<CapturedExportedSpan> capturedExportedSpans = getCapturedExportedSpansWithTraceState();
+
+    SpanTestHierarchy spanTestHierarchy = new SpanTestHierarchy(capturedExportedSpans);
+    spanTestHierarchy.withRoot(EXPECTED_FLOW_SPAN_NAME)
+        .addTraceStateKeyValueAssertion("key1", "value1")
+        .addTraceStateKeyValueAssertion("key2", "value2")
+        .addTraceStateKeyValueAssertion("key3", "value3")
+        .beginChildren()
+        .child(EXPECTED_ASYNC_SPAN_NAME)
+        .addTraceStateKeyValueAssertion("key1", "value1")
+        .addTraceStateKeyValueAssertion("key2", "value2")
+        .addTraceStateKeyValueAssertion("key3", "value3")
+        .beginChildren()
+        .child(EXPECTED_LOGGER_SPAN_NAME)
+        .addTraceStateKeyValueAssertion("key1", "value1")
+        .addTraceStateKeyValueAssertion("key2", "value2")
+        .addTraceStateKeyValueAssertion("key3", "value3")
+        .endChildren()
+        .endChildren();
+
+    spanTestHierarchy.assertSpanTree();
+  }
+
+  @NotNull
```

Comment:
Is this annotation necessary?

Reply:
Just for the IDE in some cases. I'll remove it.